src/libraries/System.Linq/src/System/Linq/Where.SizeOpt.cs

```diff
@@ -0,0 +1,114 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Generic;
+using System.Diagnostics;
+
+namespace System.Linq
+{
+    public static partial class Enumerable
+    {
+        private sealed partial class SizeOptIListWhereIterator<TSource> : Iterator<TSource>
+        {
+            private readonly IList<TSource> _source;
+            private readonly Func<TSource, bool> _predicate;
+            private IEnumerator<TSource>? _enumerator;
+
+            public SizeOptIListWhereIterator(IList<TSource> source, Func<TSource, bool> predicate)
+            {
+                Debug.Assert(source is not null && source.Count > 0);
```

Comment:
I guess the code of class `SizeOptIListWhereIterator` was copied from `ListWhereIterator`, but here an extra assert of `source.Count > 0` was added? Empty IList is not filtered out by caller.